tests/__init__.py

```diff
@@ -85,11 +85,13 @@ def collect_events(stop):  # noqa: 306
                                         args=(stop_event_thread,))
         event_thread.start()
 
-        callback(device)
+        try:
+            callback(device)
 
-        sleep(wait)
-        stop_event_thread.set()
-        event_thread.join()
+            sleep(wait)
```

Comment:
I *think* you should better have:

```python
try:
  callback(device)
except Exception:
  pass
else:
  sleep(wait)
finally:
  stop_event_thread.set()
  event_thread.join()
```

And a couple more questions:
- why is the `sleep(wait)` required? (1 second of wait seems long, you should be there only after having sent the whole batch of events, so just wait for the processing of the last few reports)
- can't we also have the evdev reading Thread as a fixture?

Reply:
Why are we suppressing all exceptions? I think we want to throw all unexpected exceptions, no? In fact, we have a test to make sure we throw an exception if the request data is invalid (ex. try to move 10cm in 0.2s at 800dpi and 100Hz: the total report count is not enough for the movement requested) instead of just failing silently and having random tests fail. Does that make sense?

Reply:
> why is the sleep(wait) required? (1 second of wait seems long, you should be there only after having sent the whole batch of events, so just wait for the processing of the last few reports)

1sec is the default, it can be removed. The only time we don't use the default duration is for the test that is expected an exception for all other cases we use the duration of the action.

> can't we also have the evdev reading Thread as a fixture?

No because we need to do setup first: add the actuators, set report rate, etc. We also have to define the action. Fixtures are run before the test.

Reply:
> > why is the sleep(wait) required? (1 second of wait seems long, you should be there only after having sent the whole batch of events, so just wait for the processing of the last few reports)
> 
> 1sec is the default, it can be removed. The only time we use the default duration is for the test expects an exception, for all other cases we use the duration of the action.

Hmm, so if you need to do a test that moves a mouse for 5 seconds, you will in addition to the callback wait 5 more seconds?

I am still not quite sure how the callback integrates in this function.

> 
> > can't we also have the evdev reading Thread as a fixture?
> 
> No, because we need to do setup first: add the actuators, set report rate, etc. We also have to define the action. Fixtures are run before the test.

Hmm, OK. but you should probably make this a part of the `Endpoint`: whenever you create the endpoint, you open the evdev nodes, start listening to them in the Thread attached to the endpoint, and then your test can be more reliable and simple.

Reply:
> Hmm, so if you need to do a test that moves a mouse for 5 seconds, you will in addition to the callback wait 5 more seconds?
>
> I am still not quite sure how the callback integrates in this function.

Yes, if the action takes 5 seconds, we need to wait 5 seconds. The action is performed in real-time, we need to listen to the action while it is being performed, which takes 5 seconds.

> Hmm, OK. but you should probably make this a part of the Endpoint: whenever you create the endpoint, you open the evdev nodes, start listening to them in the Thread attached to the endpoint, and then your test can be more reliable and simple.

You still need to explicitly perform the action, so the tests would look the same. The only difference is that instead of starting the read thread in `simulate`, we would do it in the in `Endpoint`.

It is also not a great design decision in my opinion. You will only need to listen to the input events sometimes. If we create 10 devices with 3 endpoints we would be creating 30 listening threads. This not seem like a great approach when you only need to get the input events for 2 or 3 devices. The listening thread would also be present all the time, instead of just while we need them. What do you think?

Reply:
> > Hmm, so if you need to do a test that moves a mouse for 5 seconds, you will in addition to the callback wait 5 more seconds?
> > I am still not quite sure how the callback integrates in this function.
> 
> Yes, if the action takes 5 seconds, we need to wait 5 seconds. The action is performed in real-time, we need to listen to the action while it is being performed, which takes 5 seconds.

Oh. Actually, I think what I missed is that the `callback` is starting a separate thread. So we better just have a synchronous `callback` call instead, given that the evdev events are already fetched in their own thread.


> 
> > Hmm, OK. but you should probably make this a part of the Endpoint: whenever you create the endpoint, you open the evdev nodes, start listening to them in the Thread attached to the endpoint, and then your test can be more reliable and simple.
> 
> You still need to explicitly perform the action, so the tests would look the same. The only difference is that instead of starting the read thread in `simulate`, we would do it in the in `Endpoint`.

Nope, see #42

The tests are now:
```
self.simulate(something)
wait()
self.catch_evdev_events()
```
Which is much more readable than the current `callback` solution where we go back and forth.

> 
> It is also not a great design decision in my opinion. You will only need to listen to the input events sometimes. If we create 10 devices with 3 endpoints we would be creating 30 listening threads. This not seem like a great approach when you only need to get the input events for 2 or 3 devices. The listening thread would also be present all the time, instead of just while we need them. What do you think?

#42 creates only one thread per Device. That is something that can be changed, for sure, but it should work for the time being.
Then, listening to evdev nodes is *cheap*. It seems massively CPU consuming right now because the evdev thread in this PR does active polling on the file descriptors. As soon as you limit the rate, you realize that they are nothing.

And if you really don't want to have the thread all the time ready you can always have something like:
```python
device.start_listening_to_evdev_events()
device.simulate_synchronously()
time.sleep(0.1) # give a little bit of time to process the incoming reports by the kernel and evdev
device.stop_listening_to_evdev_events()
device.catch_evdev_events()
```

And of course, if you are worried about opening too many file descriptors at once, you should delay the creation of the kernel device, which is what I have to do in `hid-tools` when some of the tests should be skipped based on the static definition of the report descriptors.

Reply:
> Nope, see #42

Okay, I understand your implementation proposal and the way it is implemented could be accomplished with a fixture. Like you said:

> whenever you create the endpoint, you open the evdev nodes, start listening to them in the Thread attached to the endpoint, and then your test can be more reliable and simple.

You would have a fixture that would start listening to events when the test starts and stop when the test ends.

This was misunderstanding from my part, I should have understood what you meant the first time you described it. Sorry.

Reply:
I pushed a new commit implementing this. I think it's a cleaner implementation than what we have in #42. Please have a look.

Reply:
@bentiss friendly ping, it has been over a week. Is my approach fine or do you still want to do it in `Device`?